### PR TITLE
fix(pacquet): align GVS slot layout with pnpm

### PR DESCRIPTION
**Pull request by zkochan**
## Summary

Adds three end-to-end **GVS parity tests** under `pacquet/crates/cli/tests/pnpm_compatibility.rs` that run `pnpm install` and `pacquet install --frozen-lockfile` against the same workspace + lockfile with `enableGlobalVirtualStore: true`, then diff the resulting `<store>/v11/links/` slot trees. The tests surfaced three independent divergences, each fixed in its own commit set:

1. **`<store>/v11/links` prefix.** `getStorePath` appends `STORE_VERSION` (`v11`) to the configured `storeDir` before `extendInstallOptions.ts:352` joins `'links'` onto it, so pnpm's GVS lives at `<store>/v11/links/` — pacquet's `StoreDir::links()` was one level shallower, joining onto `self.root`. Same gap on `projects()`. Anchored both under `self.v11()` so the on-disk paths agree.

2. **GVS engine-name resolution.** `ENGINE_NAME` was computed from `process.version`, which is wrong in two cases:
   - **`@pnpm/exe` SEA bundle.** The bundle has its own embedded Node, not the `node` on PATH that runs lifecycle scripts. Two pnpm installs on the same machine (one SEA, one npm-package) therefore disagreed on the cache key, partitioning the side-effects cache and the global virtual store.
   - **`engines.runtime` / `devEngines.runtime` pin.** When a project pins a Node version, pnpm downloads that Node into `node_modules/node/` and uses it to run lifecycle scripts. But the hash still anchored to whichever Node ran pnpm itself, not to the pinned Node.

   `@pnpm/engine.runtime.system-node-version` now exports `engineName(nodeVersion?)` and `findRuntimeNodeVersion(snapshotKeys)`. The override has priority; otherwise the helper falls through to `getSystemNodeVersion()` — which already prefers shell `node --version` over `process.version` in SEA contexts — and finally to `process.version` as a last resort. `@pnpm/deps.graph-hasher`'s `calcDepState`, `calcGraphNodeHash`, and `iterateHashedGraphNodes` accept an optional `nodeVersion`. Every install-side caller (`deps.graph-builder`, `installing.deps-resolver`, `installing.deps-restorer`, `installing.deps-installer/install/link`, `building.during-install`, `building.after-install`) derives the project's pinned runtime via `findRuntimeNodeVersion` once per invocation and forwards it. The legacy `ENGINE_NAME` constant in `@pnpm/constants` is unchanged so external consumers and existing tests keep working.

   Pacquet mirrors this with `find_runtime_node_major` in `install_frozen_lockfile.rs` — it scans the lockfile's `snapshots:` map for a `node@runtime:<version>` entry and uses that major outright, only falling back to the host probe when no pin is present.

3. **Slot bin-shim layout.** Pacquet was emitting `.cmd` / `.ps1` shims on every host platform, even though pnpm only writes them on Windows ([`@zkochan/cmd-shim` `createCmdFile: isWindows`](https://github.com/pnpm/cmd-shim/blob/0d79ca9534/src/index.ts#L32) + `bins/linker`'s [`POWER_SHELL_IS_SUPPORTED = IS_WINDOWS`](https://github.com/pnpm/pnpm/blob/29a42efc3b/bins/linker/src/index.ts#L28) gate). Pacquet also excluded the slot's own package from the slot-local `node_modules/.bin/` based on a stale assumption ("which pnpm doesn't"), but pnpm's [`linkBinsOfDependencies`](https://github.com/pnpm/pnpm/blob/29a42efc3b/building/during-install/src/index.ts#L272-L298) appends `depNode` to the bin-source list unconditionally, so a leaf package like `hello-world-js-bin` writes a self-shim at `<slot>/node_modules/<pkg>/node_modules/.bin/<pkg>`. Both behaviors now match pnpm.

## Test plan

- [x] `cargo nextest run -p pacquet-cli --test pnpm_compatibility` — 5 active tests pass, 1 ignored (see below)
- [x] `cargo nextest run -p pacquet-store-dir -p pacquet-config -p pacquet-cmd-shim -p pacquet-package-manager` — 600+ tests pass after the prefix / bin-shim updates
- [x] `same_global_virtual_store_layout_pure_js` — pacquet & pnpm produce byte-identical `<store>/v11/links/` trees for `@pnpm.e2e/hello-world-js-bin-parent`
- [x] `same_global_virtual_store_layout_diamond` — same for `pkg-with-1-dep` + `parent-of-pkg-with-1-dep`, verifying `calc_dep_graph_hash` memoization parity
- [x] Three new TS unit tests in `engine/runtime/system-node-version/test/` cover the `engineName(version)` override branch and `findRuntimeNodeVersion`'s extraction rule (with and without peer suffix)
- [ ] `same_global_virtual_store_layout_with_approved_postinstall` is currently `#[ignore]`d. It requires pnpm and pacquet to agree on the `<platform>;<arch>;node<major>` triple in the engine-included hash branch. The `pnpm/setup` action on CI installs an `@pnpm/exe` SEA bundle whose embedded Node (node26) differs from the runner's PATH `node` (node24), so the digests don't line up. The pnpm-side fix in this PR resolves `engineName()` via `getSystemNodeVersion()` which prefers the shell `node`, so once a published pnpm version with the fix reaches `pnpm/setup` the test will pass without modification — re-enable it then. The other two GVS parity tests are unaffected since they exercise the engine-agnostic branch.

## Notes

- Two pacquet integration tests in `package-manager/src/install/tests.rs` had hard-coded `<store_dir>/projects/` assertions; updated to `<store_dir>/v11/projects/` to follow the prefix fix.
- The `link_bins_rewrites_when_only_sh_flavor_exists` cmd-shim test is now `#[cfg(windows)]` — the upgrade-recovery scenario it exercises is meaningless on Unix where `.cmd`/`.ps1` are no longer written in the first place.
- Review feedback addressed: (a) test YAML helper now guarantees a trailing newline before appending GVS keys; (b) `findRuntimeNodeVersion` calls in `installing/deps-restorer/` switched from `Object.keys(graph)` (install-dir-keyed in that module) to extracting `depPath` per node, with the computation lifted out of the recursion; (c) `dlx.e2e.ts`'s `jest.unstable_mockModule` against `@pnpm/engine.runtime.system-node-version` now forwards every exported symbol so transitive importers of `engineName` don't break.
- Known caveat: pacquet's non-lockfile install path (`run_with_readdir`) still excludes the slot's own bin via `link_bins_excluding`. That path runs only for the legacy flat layout where GVS parity isn't a constraint, so it's deliberately out of scope here.
- Known caveat tracked in #11690: when a dependency's own manifest declares `engines.runtime`, the resolver desugars it into a regular `dependencies.node: 'runtime:<v>'` entry on that package, so the **deps** portion of the hash captures it on both sides. The **engine** portion is still install-wide rather than per-snapshot, so cached side-effects for dep-pinned runtimes can be reused under the wrong host Node. pnpm has this same gap today; closing it on both sides requires per-snapshot engine resolution and is outside this PR's scope.

---
Written by an agent (Claude Code, claude-opus-4-7).

## Summary by CodeRabbit

* **Bug Fixes**
  * Side-effects cache keys and global-virtual-store hashing now follow project/runtime Node version (may cause one-time slot churn on upgrade).
  * Global virtual store paths use the v11 layout.

* **New Features**
  * Bin linking now includes a package’s own binaries.

* **Behavior Changes**
  * Shim generation now emits Windows (.cmd/.ps1) only on Windows and canonical shell shims on Unix.

* **Tests**
  * Added integration and unit tests validating GVS parity and runtime-sensitive hashing.

* **Documentation**
  * Config docs updated to reference v11-style store paths.

### CODE REVIEW

File: pacquet/crates/cli/tests/pnpm_compatibility.rs [hunk 1]
```diff
@@ -205,3 +205,208 @@ fn pnpm_reads_pacquet_written_rows() {
...
+
+/// Append GVS opt-in (and any extra fields) to the `pnpm-workspace.yaml`
+/// that [`CommandTempCwd::add_mocked_registry`] already populated with
+/// `storeDir` / `cacheDir`. `enableGlobalVirtualStore: true` is the
+/// switch that flips both pnpm and pacquet to the shared-store layout.
+fn enable_gvs_in_workspace_yaml(workspace: &std::path::Path, extra_yaml: &str) {
+    let yaml_path = workspace.join("pnpm-workspace.yaml");
+    let existing = fs::read_to_string(&yaml_path).expect("read pnpm-workspace.yaml");
+    let extended = format!("{existing}enableGlobalVirtualStore: true\n{extra_yaml}");
+    fs::write(&yaml_path, extended).expect("write pnpm-workspace.yaml");
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/cli/tests/pnpm_compatibility.rs]: Resolved in bb3247aba3 — `enable_gvs_in_workspace_yaml` now reads the existing YAML, pushes a `\n` if one isn't present at the tail, then appends the GVS keys. The two call sites of this helper in `pnpm_compatibility.rs` always write a trailing-newlined file in front of it today, but defending against the no-newline edge keeps the test bullet-proof against future churn in `add_mocked_registry`.

---
Written by an agent (Claude Code, claude-opus-4-7).

File: installing/deps-restorer/src/index.ts [hunk 2]
```diff
@@ -909,6 +910,11 @@ async function linkAllPkgs (
     needsBuildMarkerSrc = path.join(opts.storeDir, '.pnpm-needs-build-marker')
     await fs.writeFile(needsBuildMarkerSrc, '')
   }
+  // Resolved `engines.runtime` Node version (when present) anchors
+  // the side-effects-cache key prefix to the script-runner Node, not
+  // pnpm's own `process.version`. Computed once — the graph is
+  // closed-over by every per-node call below.
+  const nodeVersion = findRuntimeNodeVersion(Object.keys(opts.depGraph))
```
[the remark this answers is not in the document]
Comment (zkochan) [reply installing/deps-restorer/src/index.ts]: You're right — the restorer's `DependenciesGraph` is keyed by install directory, so `Object.keys(opts.depGraph)` never matched `node@runtime:<version>`. Fixed in bb3247aba3 by pulling the depPath off each node instead: `findRuntimeNodeVersion(depNodes.map((node) => node.depPath))`. Same fix applied to the sibling site in `linkHoistedModules.ts` (your other comment).

---
Written by an agent (Claude Code, claude-opus-4-7).

File: installing/deps-restorer/src/linkHoistedModules.ts [hunk 3]
```diff
@@ -102,6 +103,14 @@ async function linkAllPkgsInOrder (
...
   }
 ): Promise<void> {
+  // Pin the side-effects-cache key to the project's
+  // `engines.runtime`-resolved Node version when one is present in
+  // the graph. Otherwise the key partitions by `process.version`
+  // (pnpm's runner), and a SEA-bundled pnpm reading a cache
+  // populated by an npm-bundled pnpm would miss every entry.
+  // Computed once and shared across the recursion — the graph
+  // doesn't change while a single linker is running.
+  const nodeVersion = findRuntimeNodeVersion(Object.keys(graph))
```
[the remark this answers is not in the document]
Comment (zkochan) [reply installing/deps-restorer/src/linkHoistedModules.ts]: Both points addressed in bb3247aba3:

1. `graph` in `linkHoistedModules.ts` is keyed by install directory, so the scan now runs over `Object.values(graph).map((node) => node.depPath)` instead.
2. Lifted the computation out of the recursion — `linkHoistedModules()` resolves the runtime version once before dispatching to `linkAllPkgsInOrder`, and threads it down through `opts.nodeVersion`. The recursive call at line 171 picks it up via the spread.

---
Written by an agent (Claude Code, claude-opus-4-7).

File: pacquet/crates/cmd-shim/src/link_bins.rs [hunk 4]
```diff
@@ -425,23 +439,33 @@ where
...
-        Api::write(&ps1_path, ps1_body.as_bytes())
-            .map_err(|error| LinkBinsError::WriteShim { path: ps1_path.clone(), error })?;
+        if let Some((cmd_path, cmd_body, ps1_path, ps1_body)) = &windows_shims {
+            Api::write(cmd_path, cmd_body.as_bytes()).map_err(|error| {
+                LinkBinsError::WriteShim { path: cmd_path.clone(), error }
+            })?;
+            Api::write(ps1_path, ps1_body.as_bytes()).map_err(|error| {
+                LinkBinsError::WriteShim { path: ps1_path.clone(), error }
+            })?;
+        }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/cmd-shim/src/link_bins.rs:466]: Acknowledged — this one isn't fixed in bb3247aba3.

The convergence concern is real for a pacquet v1 → v2 upgrade-in-place on the same store: legacy `.cmd` / `.ps1` siblings written by the old all-three-flavors writer would linger in the slot forever, and a Unix install that touches the slot afterwards wouldn't repair the gap because `windows_shims is None` makes the `already_correct` check pass.

I'm leaving it out of this PR for two reasons. First, the GVS hash itself is unchanged between v1 and v2 (the engine portion is the same modulo the engine-name fix landed earlier in the PR), so a fresh install — including any `pnpm store prune` + reinstall cycle — already produces the cleaned-up slot. Second, adding active cleanup means widening the writer's capability trait bound (`FsRemoveFile` / equivalent) and threading it through every test impl, plus an error variant — non-trivial diff for an upgrade-only edge case.

Will track as a follow-up once the rest of the PR lands; happy to revisit in review if you'd prefer it in this PR.

---
Written by an agent (Claude Code, claude-opus-4-7).

**Comment by zkochan** [comment]
Filed #11690 as a follow-up for the per-snapshot engine resolution case raised in review — when a dependency declares its own `engines.runtime` pin, the resolver desugars it into a regular dep at `installing/deps-resolver/src/resolveDependencies.ts:1477-1479` and the deps half of the GVS hash captures it, but the engine field still uses the install-wide value. This PR intentionally leaves that asymmetry untouched (pnpm has it today, so pacquet matches pnpm); #11690 outlines the per-snapshot refactor needed to close it on both sides.

---
Written by an agent (Claude Code, claude-opus-4-7).

**Comment by zkochan** [comment]
@coderabbitai review
